### PR TITLE
Fix wrong line number in code search result

### LINKED ISSUE #29136: Wrong linenumber in search-result, when line is to long
### Description

When creating files with extra long lines, the text wraps and the linenumber is on the wrong line

### Gitea Version

1.21.5

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

image
Line 9 should be the line with "89485bfaa6a42365fbb052f9b34fc809e1bf6d3ed0c3b29376a22104030780c12"

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Download and systemd

### Database

SQLite

### PR DESCRIPTION
**Pull request by yp05327**
Fix #29136

Before:
The result is a table and all line numbers are all in one row.
image
image

After:
image
image

~~Updated:~~
~~added `active` class to the target line.~~
image

### CODE REVIEW

**Comment by lunny** [comment]
The map is unordered? How about using two slices?

File: templates/code/searchresults.tmpl [hunk 1]
```diff
@@ -25,14 +25,14 @@
...
-									{{range .LineNumbers}}
-										<a href="{{$repo.Link}}/src/commit/{{$result.CommitID | PathEscape}}/{{$result.Filename | PathEscapeSegments}}#L{{.}}"><span>{{.}}</span></a>
-									{{end}}
+									<a href="{{$repo.Link}}/src/commit/{{$result.CommitID | PathEscape}}/{{$result.Filename | PathEscapeSegments}}#L{{$k}}"><span>{{$k}}</span></a>
 								</td>
-								<td class="lines-code chroma"><code class="code-inner">{{.FormattedLines}}</code></td>
+								<td class="lines-code chroma"><code class="code-inner">{{$line}}</code></td>
 							</tr>
+							{{end}}
 						</tbody>
```
Comment (delvh) [inline templates/code/searchresults.tmpl]: This block sounds like it should be a subtemplate instead as it exists twice.

Comment (yp05327) [reply templates/code/searchresults.tmpl]: Done.

File: templates/shared/searchfile.tmpl [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<div class="file-body file-code code-view">
+	<table>
+		<tbody>
+			{{range .result.Lines}}
+				<tr>
+					<td class="lines-num">
+						<a href="{{$.repolink}}/src/commit/{{PathEscape $.result.CommitID}}/{{PathEscapeSegments $.result.Filename}}#L{{.Num}}"><span>{{.Num}}</span></a>
+					</td>
+					<td class="lines-code chroma"><code class="code-inner">{{.FormattedContent}}</code></td>
```
Comment (wxiaoguang) [inline templates/shared/searchfile.tmpl:9]: When you render `string` into HTML, it will be escaped. So, isn't it causing double-escaping?

Comment (yp05327) [reply templates/shared/searchfile.tmpl:9]: It happened. So need to convert it back to HTML. Now I understand why it was `template.HTML` 👍
image
image

File: templates/repo/search.tmpl [hunk 3]
```diff
@@ -44,20 +44,7 @@
...
-													{{range .LineNumbers}}
-														<a href="{{$.SourcePath}}/src/commit/{{PathEscape $result.CommitID}}/{{PathEscapeSegments $result.Filename}}#L{{.}}"><span>{{.}}</span></a>
-													{{end}}
-												</td>
-												<td class="lines-code chroma"><code class="code-inner">{{.FormattedLines}}</code></td>
-											</tr>
-										</tbody>
-									</table>
-								</div>
+								{{template "shared/searchfile" dict "repolink" $.SourcePath "result" .}}
```
Comment (wxiaoguang) [inline templates/repo/search.tmpl]: Could it follow most other `{{template}}` styles?


```suggestion
								{{template "shared/searchfile" dict "RepoLink" $.SourcePath "SearchResult" .}}
```

Comment (yp05327) [reply templates/repo/search.tmpl]: Do we have a clearly rule for this? I also confused which one to use. At last, I followed the format in L49.

image

Comment (wxiaoguang) [reply templates/repo/search.tmpl]: > Do we have a clearly rule for this? I also confused which one to use. At last, I followed the format in L49.

"root" is special ........ it means global. Some legacy code used single word and camelCase, there is no clear rule for it, but I think we should stick to Golang's style PascalCase means a public member, and two-word name makes it more maintainable.

**Review by wxiaoguang — APPROVED** [review]

**Comment by wxiaoguang** [comment]
There seems to be CI failures

**Comment by yp05327** [comment]
CI fail is caused by the file which is end with two `\n`, and we have another issue cause by this: 
image
image

**Comment by yp05327** [comment]
It is caused by the code here. We removed `\n`, and it seems that chroma didn't add newlines.
image

**Comment by silverwind** [comment]
If your cases has single `\n`, then UI should not render it, but monaco should.

**Comment by yp05327** [comment]
So `3 lines` here is expected? even there are only two lines below.
image

And we should also not render it in search results?

**Comment by yp05327** [comment]
If the content is `aaa\n\n`, when we search `aaa`, we will get `aaa\n`.
And the `\n` here is not the last `\n` of this file, so we should render it in the search result?
But if the content is `aaa\n`, when we search `aaa`, we will also get `aaa\n`.
But the `\n` here is the last `\n` in this file, so how should we deal with it in this case? 🤔

**Comment by silverwind** [comment]
> So `3 lines` here is expected? even there are only two lines below. image
> 
> And we should also not render it in search results?

`a\n` expectance is "1 line", no final newline in code view or search results, final newline visible in monaco.
`a\n\n` expectance is "2 lines", no final newline in code view or search results, final newline visible in monaco.

**Comment by silverwind** [comment]
> If the content is `aaa\n\n`, when we search `aaa`, we will get `aaa\n`. And the `\n` here is not the last `\n` of this file, so we should render it in the search result? But if the content is `aaa\n`, when we search `aaa`, we will also get `aaa\n`. But the `\n` here is the last `\n` in this file, so how should we deal with it in this case? 🤔

Search should just be a simple "contains" query in the raw text, including matching any trailing newline. It is only hidding during view rendering, does not affect the search.

**Comment by wxiaoguang** [comment]
image

image

**Comment by silverwind** [comment]
@yp05327 does this PR fix https://github.com/go-gitea/gitea/issues/29499?

**Comment by yp05327** [comment]
> @yp05327 does this PR fix #29499?

No.
Is that a bug? In #27386:
``` text
When rendering, the last empty line is not rendered in UI, while the line-number is still counted, to tell users that the file contains a trailing EOL.
```
Related: https://github.com/go-gitea/gitea/issues/27377#issuecomment-1742102993

**Comment by silverwind** [comment]
It's indented behaviour that final newline does not render in code view and this matches GitHub behaviour. It's just that our line count is off by 1 if there is a final newline.

**Comment by lunny** [comment]
image

**Comment by yp05327** [comment]
> image

What is your test content?

**Review by lunny — APPROVED** [review]

**Review by silverwind — APPROVED** [review]

**Pull request merged**